python/sglang/srt/managers/tp_worker.py

```diff
@@ -211,48 +220,95 @@ def __init__(
-                    recv_req, (TokenizedGenerateReqInput, TokenizedEmbeddingReqInput)
-                ):
-                    self.handle_generate_request(recv_req)
-                elif isinstance(recv_req, FlushCacheReq):
-                    self.flush_cache()
-                elif isinstance(recv_req, AbortReq):
-                    self.abort_request(recv_req)
-                elif isinstance(recv_req, UpdateWeightReqInput):
-                    success, message = self.update_weights(recv_req)
-                    self.out_pyobjs.append(UpdateWeightReqOutput(success, message))
-                else:
-                    raise ValueError(f"Invalid request: {recv_req}")
-
             # Forward
-            self.forward_step()
+            self.control_loop(recv_reqs)
         except Exception:
             logger.error("Exception in ModelTpServer:\n" + get_exception_traceback())
             raise
 
         # Return results
-        ret = self.out_pyobjs
-        self.out_pyobjs = []
+        ret = []
+        while not self.out_pyobjs_queue.empty():
+            try:
+                ret.append(self.out_pyobjs_queue.get_nowait())
+            except queue.Empty:
+                break
         return ret
 
+    def control_loop(self, recv_reqs: List):
+
+        # cache finished requests and send output to detokenizer
+        finished_requests = []
+        while not self.finished_requests.empty():
+            try:
+                finished_requests.append(self.finished_requests.get_nowait())
+            except queue.Empty:
+                break
+        self.send_to_detokenizer(finished_requests)
+        for req in finished_requests:
+            self.tree_cache.cache_finished_req(req)
+
+        # put retracted requests back to waiting queue
+        while not self.retracted_requests.empty():
+            try:
+                self.waiting_queue.append(self.retracted_requests.get_nowait())
+            except queue.Empty:
+                break
+
+        # pool new requests
+        for recv_req in recv_reqs:
+            if isinstance(
+                recv_req, (TokenizedGenerateReqInput, TokenizedEmbeddingReqInput)
+            ):
+                self.handle_generate_request(recv_req)
+            elif isinstance(recv_req, FlushCacheReq):
+                self.flush_cache()
+            elif isinstance(recv_req, AbortReq):
+                self.abort_request(recv_req)
+            elif isinstance(recv_req, UpdateWeightReqInput):
+                success, message = self.update_weights(recv_req)
+                self.out_pyobjs_queue.put(UpdateWeightReqOutput(success, message))
+            else:
+                raise ValueError(f"Invalid request: {recv_req}")
+
+        if self.phase_indicator is not Phase.PREFILLing:
+            if self.ready_prefill_batch.empty():
+                new_batch = self.get_new_prefill_batch()
+                if new_batch is not None:
+                    self.ready_prefill_batch.put(new_batch)
```

Comment:
During the multiple steps of decode, after the first time of successfully get new prefill batch, the released memory from future decode steps will not be utilized before next prefill step?

Reply:
It is not guaranteed for now, but an alternative way is to make the next prefill batch preparation to be a continuous manner, e.g., the prepared batch could be updated while the compute loop is doing decoding. This could be done by changing the schedule policy. There are some changes needed for `policy_scheduler` to make the scheduling more robust as I mentioned. I will put this change in that PR later.